Google Docs api connector

s-vamshi opened the pull request:
Resolves [#7058](https://github.com/ballerina-platform/ballerina-library/issues/7058)

## Purpose
 To introduce a new Ballerina connector for the latest Google Docs REST API by generated using it's OpenAPI specification.

## Examples

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests
- [ ] Updated the spec
- [ ] Checked native-image compatibility

s-vamshi committed:
openapi.yaml spec added

s-vamshi committed:
client generation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ballerina-platform/module-ballerinax-googleapis.gdocs?pullRequest=3) <br/>All committers have signed the CLA.

s-vamshi commented:
hey @Nuvindu, as you mentioned i have created this pr with two commits one with openapi.yaml and other with client generation.
Currently updating the test cases and would be pushing them soon!

s-vamshi committed:
wrap $ ref into allOf

https://stackoverflow.com/questions/33629750/swagger-schema-properties-ignored-when-using-ref-why

s-vamshi committed:
single missing backtick added

s-vamshi committed:
missing descriptions for request payload added

s-vamshi committed:
client generation after sanitation

Nuvindu commented on ballerina/client.bal:
```diff
@@ -13,3 +16,85 @@
+    #
+    # + headers - Headers to be sent with the request 
+    # + queries - Queries to be sent with the request 
+    # + payload - The JSON payload containing document details to be created. Only the `title` field is required and used in this request. Any other fields, including content, are ignored. 
```
Can we include these sanitations info in the `sanitations.md` file?

Nuvindu commented on ballerina/Ballerina.toml:
```diff
@@ -5,8 +5,8 @@ name = "googleapis.gdocs"
 authors = ["Ballerina"]
-keywords = []
-# icon = "icon.png" # TODO: update icon.png
+keywords = [] # TODO: Add keywords
```
Let's add some relevant keywords here?

s-vamshi replied:
Sure working on this

s-vamshi committed:
ballerina.toml file updated with keywords and icon

s-vamshi committed:
sanitations.md file updated with relevant info

s-vamshi committed:
test.bal file created and test cases added

s-vamshi commented:
hey @Nuvindu made mentioned changes and added test.bal file too, please review this in your free time!
started working on examples and package.md !

Nuvindu commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,122 @@
+// AUTO-GENERATED FILE. DO NOT MODIFY.
+// This file is auto-generated by the Ballerina OpenAPI tool.
```
```suggestion
// This file is auto-generated by the Ballerina OpenAPI tool.

```

Nuvindu commented on docs/spec/sanitations.md:
```diff
@@ -1,24 +1,24 @@
-1. 
-2. 
-3. 
+[//]: # "TODO: Add sanitation details"
```
```suggestion
```

Nuvindu commented on docs/spec/sanitations.md:
```diff
@@ -1,24 +1,24 @@
+[//]: # "TODO: Add sanitation details"
+
+1. Wrapping $ref into allOf - In OpenAPI 2.0 and 3.0.x versions, $ref works by replacing itself and all of its sibling elements with the definition it is pointing at. So the workaround was to wrap $ref into allOf. [Stackoverflow Reference](https://stackoverflow.com/questions/33629750/swagger-schema-properties-ignored-when-using-ref-why)
+2. Fix missing single backtick in sectionBreakLocation property description.
```
```suggestion
2. Fix missing single backtick in `sectionBreakLocation` property description.
```

Nuvindu commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,122 @@
+    }
+};
+
+Client gdocsClient = test:mock(Client);
```
Shall we modify this test cases to run on both mock and live endpoints? 
You can follow this twitter connector as a reference. https://github.com/ballerina-platform/module-ballerinax-twitter/tree/main/ballerina/tests

Nuvindu replied:
In order to create a mock endpoint, you may need to create a mock service. You can use the following command to generate the skeleton for the mock service.

```
bal openapi -i openapi.yaml --mode service
```

s-vamshi replied:
Sure will start working on this!

s-vamshi replied:
hey @Nuvindu
while generating a mock service using above command you mentioned i am getting following error,
![image](https://github.com/user-attachments/assets/307863e9-d2f2-4e2a-bda7-d2b39be31f2d)

Seems like : in url is not accepted by ballerina while generating service

Nuvindu replied:
Hey @s-vamshi, 
This happens since the spec has a complex resource path which is not supported in the lang spec. What you can do is comment out that complex path (From [L144](https://github.com/s-vamshi/module-ballerinax-googleapis.gdocs/blob/main/docs/spec/openapi.yaml#L144) to [L194](https://github.com/s-vamshi/module-ballerinax-googleapis.gdocs/blob/main/docs/spec/openapi.yaml#L194)) and try to generate the service skeleton.

s-vamshi replied:
hey @Nuvindu,
If i comment batchUpdate function wont be created right?
Is that ok?

Nuvindu replied:
For now, let's skip that. I will come up with a workaround for that

s-vamshi replied:
Sure!

Nuvindu replied:
@s-vamshi A kind reminder on this

s-vamshi replied:
hey @Nuvindu
After generating the mock service when I am trying to run it, I am getting an error which says Document type is not defined, even though it is defined in the types.bal

Nuvindu replied:
Where have you placed the `openapi_service.bal` file. Directly in the `tests` directory or another place?

s-vamshi replied:
Hey @Nuvindu
Placed it directly inside tests folder and renamed it to mock_service and was trying to run it using bal run tests/mock_service.bal

Nuvindu replied:
@s-vamshi , If you run it as `bal run tests/mock_service.bal` it will be executed as a single Ballerina file. Then the types from other .bal files won't be recognized. Instead, go to the `/ballerina` directory and execute the `bal test` command. This will run all services and test cases within the project.

Nuvindu replied:
Hi @s-vamshi, 
Let me know once the task is completed .

s-vamshi commented:
hey @Nuvindu
while generating a mock service using above command you mentioned i am getting following error,
![image](https://github.com/user-attachments/assets/307863e9-d2f2-4e2a-bda7-d2b39be31f2d)

Seems like : in url is not accepted by ballerina while generating service

Nuvindu commented:
Closed this due to inactivity. We can reopen the pull request once the contributor is ready to work on this
